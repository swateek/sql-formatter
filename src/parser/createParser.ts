import { Parser as NearleyParser, Grammar } from 'nearley';

import Tokenizer from 'src/lexer/Tokenizer';
import { disambiguateTokens } from 'src/lexer/disambiguateTokens';
import { ParamTypes } from 'src/lexer/TokenizerOptions';
import { StatementNode } from 'src/parser/ast';
import grammar from 'src/parser/grammar';
import LexerAdapter from 'src/parser/LexerAdapter';
import { EOF_TOKEN } from 'src/lexer/token';

export interface Parser {
  parse(sql: string, paramTypesOverrides: ParamTypes): StatementNode[];
}

/**
 * Creates a parser object which wraps the setup of Nearley parser
 */
export function createParser(tokenizer: Tokenizer): Parser {
  let paramTypesOverrides: ParamTypes = {};
  const lexer = new LexerAdapter(chunk => [
    ...disambiguateTokens(tokenizer.tokenize(chunk, paramTypesOverrides)),
    EOF_TOKEN,
  ]);
  const parser = new NearleyParser(Grammar.fromCompiled(grammar), { lexer });

  return {
    parse: (sql: string, paramTypes: ParamTypes) => {
      // share paramTypesOverrides with Tokenizer
      paramTypesOverrides = paramTypes;

      const { results } = parser.feed(sql);

      if (results.length === 1) {
        return results[0];
      } else if (results.length === 0) {
        // Ideally we would report a line number where the parser failed,
        // but I haven't found a way to get this info from Nearley :(
        throw new Error('Parse error: Invalid SQL');
      } else {
        throw new Error('Parse error: Ambiguous grammar');
      }
    },
  };
}
