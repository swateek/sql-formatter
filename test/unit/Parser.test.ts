import Tokenizer from '../../src/lexer/Tokenizer.js';
import { createParser } from '../../src/parser/createParser.js';

describe('Parser', () => {
  const parse = (sql: string) => {
    const tokenizer = new Tokenizer({
      reservedClauses: ['FROM', 'WHERE', 'LIMIT', 'CREATE TABLE'],
      reservedSelect: ['SELECT'],
      reservedSetOperations: ['UNION', 'UNION ALL'],
      reservedJoins: ['JOIN'],
      reservedFunctionNames: ['SQRT', 'CURRENT_TIME'],
      reservedKeywords: ['BETWEEN', 'LIKE', 'ON', 'USING'],
      operators: [':'],
      extraParens: ['[]', '{}'],
      stringTypes: ["''-qq"],
      identTypes: ['""-qq'],
    });

    return createParser(tokenizer).parse(sql, {});
  };

  it('parses empty list of tokens', () => {
    expect(parse('')).toEqual([]);
  });

  it('throws error when parsing invalid SQL expression', () => {
    expect(() => parse('SELECT (')).toThrow('Parse error at token: «EOF»');
  });

  it('parses list of statements', () => {
    expect(parse('foo; bar')).toMatchSnapshot();
  });

  it('parses array subscript', () => {
    expect(parse('SELECT my_array[5]')).toMatchSnapshot();
  });

  it('parses array subscript with comment', () => {
    expect(parse('SELECT my_array /*haha*/ [5]')).toMatchSnapshot();
  });

  it('parses parenthesized expressions', () => {
    expect(parse('SELECT (birth_year - (CURRENT_DATE + 1))')).toMatchSnapshot();
  });

  it('parses function call', () => {
    expect(parse('SELECT sqrt(2)')).toMatchSnapshot();
  });

  it('parses LIMIT clause with count', () => {
    expect(parse('LIMIT 15;')).toMatchSnapshot();
  });

  it('parses LIMIT clause with offset and count', () => {
    expect(parse('LIMIT 100, 15;')).toMatchSnapshot();
  });

  it('parses LIMIT clause with longer expressions', () => {
    expect(parse('LIMIT 50 + 50, 3 * 2;')).toMatchSnapshot();
  });

  it('parses BETWEEN expression', () => {
    expect(parse('WHERE age BETWEEN 18 AND 63')).toMatchSnapshot();
  });

  it('parses set operations', () => {
    expect(parse('SELECT foo FROM bar UNION ALL SELECT foo FROM baz')).toMatchSnapshot();
  });

  it('parses SELECT *', () => {
    expect(parse('SELECT *')).toMatchSnapshot();
  });

  it('parses SELECT ident.*', () => {
    expect(parse('SELECT ident.*')).toMatchSnapshot();
  });

  it('parses function name with and without parameters', () => {
    expect(parse('SELECT CURRENT_TIME a, CURRENT_TIME() b;')).toMatchSnapshot();
  });

  it('parses curly braces', () => {
    expect(parse('SELECT {foo: bar};')).toMatchSnapshot();
  });

  it('parses square brackets', () => {
    expect(parse('SELECT [1, 2, 3];')).toMatchSnapshot();
  });

  it('parses qualified.identifier.sequence', () => {
    expect(parse('SELECT foo.bar.baz;')).toMatchSnapshot();
  });

  it('parses CASE expression', () => {
    expect(parse('SELECT CASE foo WHEN 1+1 THEN 10 ELSE 20 END;')).toMatchSnapshot();
  });
});
