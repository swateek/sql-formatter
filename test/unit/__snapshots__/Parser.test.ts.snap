// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parses BETWEEN expression 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "age",
            "type": "identifier",
          },
          Object {
            "and": Object {
              "raw": "AND",
              "text": "AND",
              "tokenType": "AND",
              "type": "keyword",
            },
            "between": Object {
              "raw": "BETWEEN",
              "text": "BETWEEN",
              "tokenType": "BETWEEN",
              "type": "keyword",
            },
            "expr1": Array [
              Object {
                "text": "18",
                "type": "literal",
              },
            ],
            "expr2": Array [
              Object {
                "text": "63",
                "type": "literal",
              },
            ],
            "type": "between_predicate",
          },
        ],
        "name": Object {
          "raw": "WHERE",
          "text": "WHERE",
          "tokenType": "RESERVED_COMMAND",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses LIMIT clause with count 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "count": Array [
          Object {
            "text": "15",
            "type": "literal",
          },
        ],
        "name": Object {
          "raw": "LIMIT",
          "text": "LIMIT",
          "tokenType": "LIMIT",
          "type": "keyword",
        },
        "type": "limit_clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses LIMIT clause with longer expressions 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "count": Array [
          Object {
            "text": "3",
            "type": "literal",
          },
          Object {
            "text": "*",
            "type": "operator",
          },
          Object {
            "text": "2",
            "type": "literal",
          },
        ],
        "name": Object {
          "raw": "LIMIT",
          "text": "LIMIT",
          "tokenType": "LIMIT",
          "type": "keyword",
        },
        "offset": Array [
          Object {
            "text": "50",
            "type": "literal",
          },
          Object {
            "text": "+",
            "type": "operator",
          },
          Object {
            "text": "50",
            "type": "literal",
          },
        ],
        "type": "limit_clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses LIMIT clause with offset and count 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "count": Array [
          Object {
            "text": "15",
            "type": "literal",
          },
        ],
        "name": Object {
          "raw": "LIMIT",
          "text": "LIMIT",
          "tokenType": "LIMIT",
          "type": "keyword",
        },
        "offset": Array [
          Object {
            "text": "100",
            "type": "literal",
          },
        ],
        "type": "limit_clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses SELECT * 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "all_columns_asterisk",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses array subscript 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "array": Object {
              "text": "my_array",
              "type": "identifier",
            },
            "parenthesis": Object {
              "children": Array [
                Object {
                  "text": "5",
                  "type": "literal",
                },
              ],
              "closeParen": "]",
              "openParen": "[",
              "type": "parenthesis",
            },
            "type": "array_subscript",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses curly braces 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "foo",
                "type": "identifier",
              },
              Object {
                "text": ":",
                "type": "operator",
              },
              Object {
                "text": "bar",
                "type": "identifier",
              },
            ],
            "closeParen": "}",
            "openParen": "{",
            "type": "parenthesis",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses function call 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "name": Object {
              "raw": "sqrt",
              "text": "SQRT",
              "tokenType": "RESERVED_FUNCTION_NAME",
              "type": "keyword",
            },
            "parenthesis": Object {
              "children": Array [
                Object {
                  "text": "2",
                  "type": "literal",
                },
              ],
              "closeParen": ")",
              "openParen": "(",
              "type": "parenthesis",
            },
            "type": "function_call",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses function name with and without parameters 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "raw": "CURRENT_TIME",
            "text": "CURRENT_TIME",
            "tokenType": "RESERVED_KEYWORD",
            "type": "keyword",
          },
          Object {
            "text": "a",
            "type": "identifier",
          },
          Object {
            "type": "comma",
          },
          Object {
            "name": Object {
              "raw": "CURRENT_TIME",
              "text": "CURRENT_TIME",
              "tokenType": "RESERVED_FUNCTION_NAME",
              "type": "keyword",
            },
            "parenthesis": Object {
              "children": Array [],
              "closeParen": ")",
              "openParen": "(",
              "type": "parenthesis",
            },
            "type": "function_call",
          },
          Object {
            "text": "b",
            "type": "identifier",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;

exports[`Parser parses list of statements 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "foo",
        "type": "identifier",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
  Object {
    "children": Array [
      Object {
        "text": "bar",
        "type": "identifier",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses parenthesized expressions 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "birth_year",
                "type": "identifier",
              },
              Object {
                "text": "-",
                "type": "operator",
              },
              Object {
                "children": Array [
                  Object {
                    "text": "CURRENT_DATE",
                    "type": "identifier",
                  },
                  Object {
                    "text": "+",
                    "type": "operator",
                  },
                  Object {
                    "text": "1",
                    "type": "literal",
                  },
                ],
                "closeParen": ")",
                "openParen": "(",
                "type": "parenthesis",
              },
            ],
            "closeParen": ")",
            "openParen": "(",
            "type": "parenthesis",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses set operations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "foo",
            "type": "identifier",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
      Object {
        "children": Array [
          Object {
            "text": "bar",
            "type": "identifier",
          },
        ],
        "name": Object {
          "raw": "FROM",
          "text": "FROM",
          "tokenType": "RESERVED_COMMAND",
          "type": "keyword",
        },
        "type": "clause",
      },
      Object {
        "children": Array [],
        "name": Object {
          "raw": "UNION ALL",
          "text": "UNION ALL",
          "tokenType": "RESERVED_SET_OPERATION",
          "type": "keyword",
        },
        "type": "set_operation",
      },
      Object {
        "children": Array [
          Object {
            "text": "foo",
            "type": "identifier",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
      Object {
        "children": Array [
          Object {
            "text": "baz",
            "type": "identifier",
          },
        ],
        "name": Object {
          "raw": "FROM",
          "text": "FROM",
          "tokenType": "RESERVED_COMMAND",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": false,
    "type": "statement",
  },
]
`;

exports[`Parser parses square brackets 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "1",
                "type": "literal",
              },
              Object {
                "type": "comma",
              },
              Object {
                "text": "2",
                "type": "literal",
              },
              Object {
                "type": "comma",
              },
              Object {
                "text": "3",
                "type": "literal",
              },
            ],
            "closeParen": "]",
            "openParen": "[",
            "type": "parenthesis",
          },
        ],
        "name": Object {
          "raw": "SELECT",
          "text": "SELECT",
          "tokenType": "RESERVED_SELECT",
          "type": "keyword",
        },
        "type": "clause",
      },
    ],
    "hasSemicolon": true,
    "type": "statement",
  },
]
`;
